shared_model/builders/protobuf/transaction.hpp

```diff
@@ -46,131 +47,171 @@ namespace shared_model {
+
+      TemplateTransactionBuilder(const ProtoTx &tx,
+                                 const SV &stateless_validator)
+          : transaction_(tx), stateless_validator_(stateless_validator) {}
+
+      /**
+       * Make transformation on copied content
+       * @param f - transform function for proto object
+       * @return new builder with updated state
+       */
+      template <int Fields>
+      NextBuilder<Fields> transform(std::function<void(ProtoTx &)> f) const {
+        auto copy = transaction_;
+        f(copy);
+        return {copy, stateless_validator_};
+      }
 
-      SV stateless_validator_;
+      /**
+       * Make add command transformation on copied object
+       * @param f - transform function for proto command
+       * @return new builder with added command
+       */
+      NextBuilder<Command> addCommand(
+          std::function<void(ProtoCommand *)> f) const {
+        auto copy = transaction_;
+        f(copy.mutable_payload()->add_commands());
+        return {copy, stateless_validator_};
+      }
 
      public:
       TemplateTransactionBuilder() = default;
 
       NextBuilder<CreatorAccountId> creatorAccountId(
-          const interface::types::AccountIdType &account_id) {
-        transaction_.mutable_payload()->set_creator_account_id(account_id);
-        return *this;
+          const interface::types::AccountIdType &account_id) const {
+        return transform<CreatorAccountId>([&](auto &tx) {
+          tx.mutable_payload()->set_creator_account_id(account_id);
+        });
       }
 
-      NextBuilder<TxCounter> txCounter(Transaction::TxCounterType tx_counter) {
-        transaction_.mutable_payload()->set_tx_counter(tx_counter);
-        return *this;
+      NextBuilder<TxCounter> txCounter(
+          Transaction::TxCounterType tx_counter) const {
+        return transform<CreatorAccountId>([&](auto &tx) {
+          tx.mutable_payload()->set_tx_counter(tx_counter);
+        });
       }
 
       NextBuilder<CreatedTime> createdTime(
-          interface::types::TimestampType created_time) {
-        transaction_.mutable_payload()->set_created_time(created_time);
-        return *this;
+          interface::types::TimestampType created_time) const {
+        return transform<CreatorAccountId>([&](auto &tx) {
+          tx.mutable_payload()->set_created_time(created_time);
+        });
       }
 
       NextBuilder<Command> assetQuantity(
           const interface::types::AccountIdType &account_id,
           const interface::types::AssetIdType &asset_id,
-          const std::string &amount) {
-        auto command = proto_command()->mutable_add_asset_quantity();
-        command->set_account_id(account_id);
-        command->set_asset_id(asset_id);
-        addAmount(command->mutable_amount(), amount);
-        return *this;
+          const std::string &amount) const {
+        return addCommand([&](auto command) {
+          auto cmd = command->mutable_add_asset_quantity();
```

Comment:
to be consistent with other commands could be renamed: command->proto_command, cmd->command